core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1901,12 +1933,34 @@ public PreSchedulingTaskContext getContext(ScheduledTask task)
 
             public void addContext(ScheduledTask task, NodeLease nodeLease, TaskExecutionClass executionClass)
             {
-                contexts.put(task, new PreSchedulingTaskContext(nodeLease, executionClass));
+                verify(!contexts.containsKey(task), "context already present for %s", task);
+                PreSchedulingTaskContext context = new PreSchedulingTaskContext(nodeLease, executionClass);
+                contexts.put(task, context);
+                waitCounters.get(context.getExecutionClass()).incrementAndGet();
             }
 
             public PreSchedulingTaskContext removeContext(ScheduledTask task)
             {
-                return contexts.remove(task);
+                PreSchedulingTaskContext context = contexts.remove(task);
+                verify(context != null, "expected %s in preSchedulingTaskContexts", task);
+                if (context.isWaitingForNode()) {
+                    // context removed while node still being acquired. Can happen if we decide to kill stage.
```

Comment:
The context was removed while the node was still being acquired. This can happen if we decide to kill the stage.